### PR TITLE
Pool Refactor - Part 1 - Move tranche related structures

### PR DESCRIPTION
**Pull request by mustermeiszer**
I will split this refactoring into multiple smaller chunks. Although, this will duplicate a little work, I hope that the reviewing process will be better and we can follow more on the changes and discuss them. 

This part does:
* Move all tranche related structures into the tranche.rs file
* Wrap the two `Vec<Tranche>` & `Vec<EpochExecutionTranche>` structs into new types
* Implement some logic like the accumulation of supplies of all pools in methods of the new types
* Remove the role approval logic from the pool api
* Adapt tests and mocks accordingly
* Makes `Order` an `OptionQuery`

### CODE REVIEW

File: pallets/pools/src/tranche.rs [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+					redeem_starts
+						.checked_mul(10u128.pow(tranche.seniority.saturating_add(1)).into())
+						.unwrap_or(u128::MAX)
+						.into(),
+				)
+			})
+			.collect()
+	}
+
+	pub fn risk_buffers(&self) -> Vec<Perquintill> {
```
Comment (offerijns) [inline pallets/pools/src/tranche.rs:232]: This should be called `min_risk_buffers`

File: pallets/pools/src/tranche.rs [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+	}
+
+	pub fn redemptions(&self) -> Vec<Balance> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.outstanding_redeem_orders)
+			.collect()
+	}
+
+	pub fn acc_redemptions(&self) -> Result<Balance, DispatchError> {
```
Comment (offerijns) [inline pallets/pools/src/tranche.rs:198]: Can we rename this to `total_outstanding_redeem_orders`?

File: pallets/pools/src/tranche.rs [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+	}
+
+	pub fn investments(&self) -> Vec<Balance> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.outstanding_invest_orders)
+			.collect()
+	}
+
+	pub fn acc_investments(&self) -> Result<Balance, DispatchError> {
```
Comment (offerijns) [inline pallets/pools/src/tranche.rs:182]: Can we rename this to `total_outstanding_invest_orders`?

File: pallets/pools/src/tranche.rs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+	pub fn supplies(&self) -> Result<Vec<Balance>, DispatchError> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.debt.checked_add(&tranche.reserve))
+			.collect::<Option<Vec<_>>>()
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn acc_supply(&self) -> Result<Balance, DispatchError> {
```
Comment (offerijns) [inline pallets/pools/src/tranche.rs:163]: Can we rename this to `total_supply`?

File: pallets/pools/src/tranche.rs [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+					.checked_add(&solution.invest_fulfillment.mul_floor(tranche.invest))
+					.and_then(|value| {
+						value.checked_sub(&solution.redeem_fulfillment.mul_floor(tranche.redeem))
+					})
+			})
+			.collect::<Option<Vec<_>>>()
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn acc_supply_with_fulfillment(
```
Comment (offerijns) [inline pallets/pools/src/tranche.rs:292]: Can we also rename the `acc_` methods here? I realize you've somewhat copied this from the previous implementation. But I don't think `acc_` is clear, I wouldn't immediately know what it means.

File: pallets/pools/src/tranche.rs [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+					redeem_starts
+						.checked_mul(10u128.pow(tranche.seniority.saturating_add(1)).into())
+						.unwrap_or(u128::MAX)
+						.into(),
+				)
+			})
+			.collect()
+	}
+
+	pub fn risk_buffers(&self) -> Vec<Perquintill> {
```
Comment (offerijns) [inline pallets/pools/src/tranche.rs:368]: Should be called `min_risk_buffers`

**Review by offerijns — COMMENTED** [review]
My background is in Solidity, where best practice is to write self-contained smart contracts that have everything inline, minimizing imports, and minimizing abstractions. So this is going away further from that, but when trying to understand the pre-refactoring code, I definitely feel that having one big file with little imports doesn't work very well with Rust/Substrate. Thus, adding abstractions to some degree I think makes sense.

I would say, one benefit of separating it out like this is that it makes it easier to test individual parts, so I think we should add significant unit tests to `tranche.rs`.

File: pallets/pools/src/lib.rs [hunk 7]
```diff
@@ -241,13 +123,6 @@ type EpochExecutionInfoOf<T> = EpochExecutionInfo<
...
-type EpochExecutionTrancheOf<T> = EpochExecutionTranche<
-	<T as Config>::Balance,
-	<T as Config>::BalanceRatio,
-	<T as Config>::TrancheWeight,
->;
-type TrancheOf<T> =
-	Tranche<<T as Config>::Balance, <T as Config>::InterestRate, <T as Config>::TrancheWeight>;
 
 #[frame_support::pallet]
 pub mod pallet {
```
Comment (offerijns) [inline pallets/pools/src/lib.rs:128]: I can't seem to comment it in the right place, but the `RoleApproved` and `RoleRevoked` events can also be removed now.

File: pallets/pools/src/lib.rs [hunk 8]
```diff
@@ -1212,72 +1018,55 @@ pub mod pallet {
...
+				.contains(&UnhealthyState::MinRiskBufferViolated)
+			{
+				let new_tranche_supplies: Vec<_> = tranches.supplies_with_fulfillment(solution)?;
+				let risk_buffers =
+					Self::calculate_risk_buffers(&new_tranche_supplies, &tranches.prices())?;
+
+				// Score: 1 / (min risk buffer - risk buffer)
+				// A higher score means the distance to the min risk buffer is smaller
+				let non_junior_tranches =
+					tranches.as_tranche_slice().split_first().unwrap().1.iter();
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pools/src/lib.rs:1031]: Yes, I wanted to see in the next part of the refactor what additional methods make sense. Will add the above.

File: pallets/pools/src/tranche.rs [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+
+/// Types alias for Tranches
+pub(super) type TranchesOf<T> =
+	Tranches<<T as Config>::Balance, <T as Config>::InterestRate, <T as Config>::TrancheWeight>;
+
+/// Types alias for Tranche
+pub(super) type TrancheOf<T> =
+	Tranche<<T as Config>::Balance, <T as Config>::InterestRate, <T as Config>::TrancheWeight>;
+
+/// Type that indicates the senority of a tranche
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/pools/src/tranche.rs:50]: Done.

File: pallets/pools/src/tranche.rs (hunk 1 above)
Comment (mustermeiszer) [reply pallets/pools/src/tranche.rs:232]: Done.

File: pallets/pools/src/lib.rs (hunk 7 above)
Comment (mustermeiszer) [reply pallets/pools/src/lib.rs:128]: Done.

File: pallets/pools/src/tranche.rs (hunk 6 above)
Comment (mustermeiszer) [reply pallets/pools/src/tranche.rs:368]: Done.

File: pallets/pools/src/tranche.rs (hunk 2 above)
Comment (mustermeiszer) [reply pallets/pools/src/tranche.rs:198]: What is the difference between the `outstanding_invest` of the normal `Tranche` and the `invest` of the `EpochExecutionTranche` fields? From a business logic perspective.

**Comment by mustermeiszer** [comment]
As this is blocking me I am merging this without review. The refactoring branch will be reviewed anyways before being merged into parchain.

**Pull request merged**